hazelcast/src/main/java/com/hazelcast/sql/impl/expression/math/RoundTruncateFunction.java

```diff
@@ -0,0 +1,217 @@
+/*
+ * Copyright (c) 2008-2020, Hazelcast, Inc. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.hazelcast.sql.impl.expression.math;
+
+import com.hazelcast.nio.ObjectDataInput;
+import com.hazelcast.nio.ObjectDataOutput;
+import com.hazelcast.nio.serialization.IdentifiedDataSerializable;
+import com.hazelcast.sql.impl.QueryException;
+import com.hazelcast.sql.impl.SqlDataSerializerHook;
+import com.hazelcast.sql.impl.expression.BiExpressionWithType;
+import com.hazelcast.sql.impl.expression.Expression;
+import com.hazelcast.sql.impl.expression.ExpressionEvalContext;
+import com.hazelcast.sql.impl.row.Row;
+import com.hazelcast.sql.impl.type.QueryDataType;
+import com.hazelcast.sql.impl.type.QueryDataTypeFamily;
+
+import java.io.IOException;
+import java.math.BigDecimal;
+import java.math.RoundingMode;
+
+/**
+ * Implementation of ROUND/TRUNCATE functions.
+ */
+public class RoundTruncateFunction<T> extends BiExpressionWithType<T> implements IdentifiedDataSerializable {
+
+    private boolean truncate;
+
+    public RoundTruncateFunction() {
+        // No-op.
+    }
+
+    private RoundTruncateFunction(Expression<?> operand1, Expression<?> operand2, QueryDataType resultType, boolean truncate) {
+        super(operand1, operand2, resultType);
+
+        this.truncate = truncate;
+    }
+
+    public static Expression<?> create(
+        Expression<?> operand1,
+        Expression<?> operand2,
+        QueryDataType resultType,
+        boolean truncate
+    ) {
+        if (operand2 == null) {
+            QueryDataType operand1Type = operand1.getType();
+
+            // No conversion is expected for non-fractional types when the length operand is not defined.
+            if (MathFunctionUtils.notFractional(operand1Type) && operand1Type == resultType) {
+                return operand1;
```

Comment:
We can also do this if `operand2` is a constant expression that is >=0

Reply:
This is not trivial to implement, because the constant is often wrapped into a `CAST` operator due to type coercion.

Reply:
Could you please give some examples of queries producing such literals wrapped into `CAST`? Usually, a literal cast should be either unnecessary or evaluated statically. If we are frequently producing such casts at runtime, that should be considered as a potential performance issue.